### ISSUE TITLE
Coupling random user_agent (scrapy_fake_useragent) extension with  cfscrape

### ISSUE THREAD

**Issue by reyman** [Anorov/cloudflare-scrape#88]
I try to make a scrapy script which use `cfscrape`, `privoxy`, and `scrapy_fake_useragent`, but i have some difficulties to understand how i can pass the random `user_agent` generated by `scrapy_fake_useragent` extension to your great `cfscrape` extension :)

I crosspost this issue/question on `scrapy-fake-useragent` issue and `stackOverflow` with more explanation because i think other person can be interested by this coupling :

- http://stackoverflow.com/questions/41589391/share-user-agent-between-scrapy-fake-useragent-and-cfscrape-scrapy-extension

- https://github.com/alecxe/scrapy-fake-useragent/issues/9

**Comment by Anorov** [comment]
@reyman You can pass a `user_agent` keyword argument to `cfscrape.get_tokens()`.

**Comment by reyman** [comment]
@Anorov This is the first thing i try, but `scrapy_fake_useragent` is a middleware extension, so i don't know how can i catch this useragent value generated by these extension during the lifecycle of scrapy. When i try to get using `self.settings['USER_AGENT']` i have no value at this point...

```
    def start_requests(self):
        cf_requests = []
        user_agent=self.settings['USER_AGENT']
        for url in self.start_urls:
            token , agent = cfscrape.get_tokens(url,user_agent)
            self.logger.info("token = %s", token)
            self.logger.info("agent = %s", agent)

            cf_requests.append(scrapy.Request(url=url,
                                              cookies= token,
                                              headers={'User-Agent': agent}))
        return cf_requests
```

As you can see, value differ : 

```
2017-01-12 09:21:48 [airports] INFO: agent = Scrapy/1.3.0 (+http://scrapy.org)
...
2017-01-12 09:21:48 [scrapy_fake_useragent.middleware] DEBUG: Assign User-Agent Mozilla/5.0 (Windows NT 5.1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/41.0.2224.3 Safari/537.36 to Proxy None

```

**Comment by Xonshiz** [comment]
You could always make a `request` session and start a `cfscrape` instance in it and then pass the headers to requests, rather than passing to `cfscrape`.

**Comment by reyman** [comment]
@Xonshiz When you say request session, you refer to this [lib](http://docs.python-requests.org/en/master/user/advanced/) ?

**Comment by Xonshiz** [comment]
Yep. Exactly that one. You can [check here how to do what I mentioned above](https://github.com/Xonshiz/comic-dl/blob/master/comic_dl/sites/kisscomicus.py#L14)

You can do something like this : 
```

headers = {
        'User-Agent': 'Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/54.0.2840.71 Safari/537.36'
        
    }

sess = requests.session()
sess = cfscrape.create_scraper(sess)
s = sess.get("somesite.com", headers=headers)

```

This should work. Check [this](https://github.com/Xonshiz/comic-dl/blob/master/comic_dl/sites/gomanga.py#L20) for the header reference.

By mixing the two, you get the combined power of cfscrape and requests. Besides, cfscrape is build around requests lib. Read the first line of cfscrape's readme.md

> A simple Python module to bypass Cloudflare's anti-bot page (also known as "I'm Under Attack Mode", or IUAM), implemented with Requests.

**Comment by reyman** [comment]
Thanks @Xonshiz it's seems this is a good alternative solution, but i need to rewrite all my scrapy code i suppose :-/ And i lost the capacity offered by `scrapy_fake_useragent`. If ask the question to dev of the extension to know if it's possible to catch the random user_agent value generated before using it into cfscrape on project issue page.
